src/http_server/api/v1/health.c

```diff
@@ -0,0 +1,212 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_input.h>
+#include <fluent-bit/flb_filter.h>
+#include <fluent-bit/flb_log.h>
+#include <fluent-bit/flb_output.h>
+#include <fluent-bit/flb_sds.h>
+#include <fluent-bit/flb_version.h>
+#include <fluent-bit/flb_http_server.h>
+#include <msgpack.h>
+
+#include "health.h"
+
+struct flb_health_check_metrics_counter *metrics_counter;
+struct flb_health_check_status_counter *status_counter;
+struct flb_hs *all_hs;
+/* initialize the two counters */
+static void counter_init(struct flb_hs *hs) {
+
+   metrics_counter = flb_malloc(sizeof(struct flb_health_check_metrics_counter));
+
+   if (!metrics_counter) {
+       flb_errno();
+       return;
+   }
+
+    metrics_counter->error_counter = 0;
+    metrics_counter->retry_failure_counter = 0;
+    metrics_counter->error_limit = hs->config->error_counts;
+    metrics_counter->retry_failure_limit = hs->config->retry_failure_count;
+    metrics_counter->interval_counter = 0;
+    metrics_counter->interval_limit = hs->config->interval_by_second;
+
+    status_counter = flb_malloc(sizeof(struct flb_health_check_status_counter));
+    if (!metrics_counter) {
+        flb_errno();
+        return;
+    }
+
+    status_counter->healthy_datapoint_limit = hs->config->datapoint_before_recovery;
+    status_counter->healthy_counter = hs->config->datapoint_before_recovery;
+    status_counter->unhealthy_datapoint_limit = hs->config->datapoint_before_unhealthy;
+    status_counter->unhealthy_counter = 0;
+
+}
+
+/* record healthy status based on the metrics*/
+static void record_healthy_status(int error_count, int retry_failure_count)
+{
+    if (error_count >= metrics_counter->error_limit ||
+            retry_failure_count >= metrics_counter->retry_failure_limit) {
+        status_counter->unhealthy_counter++;
+        if (status_counter->unhealthy_counter >= status_counter->unhealthy_datapoint_limit) {
+            status_counter->healthy_counter = 0;
+        }
+    } else {
+        status_counter->healthy_counter++;
+        if (status_counter->healthy_counter >= status_counter->healthy_datapoint_limit) {
+            status_counter->unhealthy_counter = 0;
+        }
+    }
+}
+
+/* tell what's the current status for health check*/
+static int tell_health_status() {
+    if (status_counter->healthy_counter >= status_counter->healthy_datapoint_limit) {
+        return 1;
+    }
+    else if (status_counter->unhealthy_counter >= status_counter->unhealthy_datapoint_limit) {
+        return 0;
+    }
+    return -1;
```

Comment:
why return -1? Shouldn't the status always be healthy or unhealthy? This function should be returning a boolean basically, right?

Reply:
refactor this function, the logic change to based on the error number and period to tell the healthy status.